src/mito2/src/gc.rs

```diff
@@ -0,0 +1,420 @@
+                    Ok(e) => {
+                        if let Some(end) = &end {
+                            // reach end, stop listing
+                            e.name() < end.as_str()
+                        } else {
+                            // no end, take all entries
+                            true
+                        }
+                    }
+                    // entry went wrong, log and skip it
+                    Err(err) => {
+                        warn!("Failed to list entry: {}", err);
+                        true
+                    }
+                });
+                let stream = stream.collect::<Vec<_>>().await;
+                // ordering of files doesn't matter here, so we can send them directly
+                tx.send(stream).await.expect("Failed to send entries");
+            });
+
+            handles.push(handle);
+        }
+        // Wait for all listers to finish
+        for handle in handles {
+            handle.await.context(JoinSnafu)?;
+        }
+
+        drop(tx); // Close the channel to stop receiving
+
+        // Collect all entries from the channel
+        let mut entries = vec![];
+        while let Some(stream) = rx.recv().await {
+            entries.extend(
+                stream
+                    .into_iter()
+                    .filter_map(Result::ok)
+                    .filter(|e| !in_use_filenames.contains(e.name()))
+                    .filter(|e| !may_linger_filenames.contains(e.name()))
+                    .filter(|e| {
+                        if !all_files_appear_in_delta_manifests.contains(e.name()) {
+                            // if the file's expel time is unknown(because not appear in delta manifest), we keep it for a while
+                            // using it's last modified time
+                            // notice unknown files use a different lingering time
+                            e.metadata()
+                                .last_modified()
+                                .map(|t| t < unknown_file_may_linger_until)
+                                .unwrap_or(false)
+                        } else {
+                            // if the file did appear in manifest delta(and passes previous predicate), we can delete it immediately
+                            true
+                        }
+                    }),
+            );
+        }
+
+        Ok(entries)
+    }
+}
+
+/// Generate partition prefixes based on concurrency and
+/// assume file names are evenly-distributed uuid string,
+/// to evenly distribute files across partitions.
+/// For example, if concurrency is 2, partition prefixes will be:
+/// ["8"] so it divide uuids into two partitions based on the first character.
+/// If concurrency is 32, partition prefixes will be:
+/// ["08", "10", "18", "20", "28", "30", "38" ..., "f0", "f8"]
+/// if concurrency is 1, it returns an empty vector.
+///
+fn gen_partition_from_concurrency(concurrency: usize) -> Vec<String> {
```

Comment:
Since we may have multiple regions, maybe we should limit the max concurrency per gc job. We can also use a fixed concurrency for simplicity.